service/dap/server.go

```diff
@@ -686,16 +689,212 @@ func (s *Server) onStackTraceRequest(request *dap.StackTraceRequest) {
+	// https://github.com/go-delve/delve/blob/master/Documentation/api/ClientHowto.md#looking-into-variables
+
+	switch v.Kind {
+	case reflect.Map:
+		kvIndex := 0
+		for i := range v.Children {
+			// A map will have twice as many children as there are elements.
+			// Process children in pairs: even indices are map keys, odd indices are values.
+			if i%2 == 0 {
+				kvIndex = i / 2
+				continue
+			}
+			key, keyref := s.convertVariable(v.Children[i-1])
+			val, valref := s.convertVariable(v.Children[i])
+			// If key or value or both are scalars, we can use
+			// a single variable to represet key:value format.
+			// Otherwise, we must return separate variables for both.
+			if keyref > 0 && valref > 0 { // Both are not scalars
+				keyvar := dap.Variable{
+					Name:               fmt.Sprintf("[key %d]", kvIndex),
+					Value:              key,
+					VariablesReference: keyref,
+				}
+				valvar := dap.Variable{
+					Name:               fmt.Sprintf("[val %d]", kvIndex),
+					Value:              val,
+					VariablesReference: valref,
+				}
+				children = append(children, keyvar, valvar)
+			} else { // At least one is a scalar
+				kvvar := dap.Variable{
+					Name:               key,
+					Value:              val,
+					VariablesReference: valref,
+				}
+				if keyref != 0 {
+					kvvar.VariablesReference = keyref
+				} else if valref != 0 {
+					kvvar.VariablesReference = valref
+				}
+				children = append(children, kvvar)
+			}
+		}
+	case reflect.Slice, reflect.Array:
+		children = make([]dap.Variable, len(v.Children))
+		for i, c := range v.Children {
+			value, varref := s.convertVariable(c)
+			children[i] = dap.Variable{
+				Name:               fmt.Sprintf("[%d]", i),
+				Value:              value,
+				VariablesReference: varref,
+			}
+		}
+	default:
+		children = make([]dap.Variable, len(v.Children))
+		for i, c := range v.Children {
+			value, variablesReference := s.convertVariable(c)
+			children[i] = dap.Variable{
+				Name:               c.Name,
+				Value:              value,
+				VariablesReference: variablesReference,
+			}
+		}
+	}
+	response := &dap.VariablesResponse{
+		Response: *newResponse(request.Request),
+		Body:     dap.VariablesResponseBody{Variables: children},
+		// TODO(polina): support evaluateName field
+	}
+	s.send(response)
+}
+
+// convertVariable converts api.Variable to dap.Variable value and reference.
+// Variable reference is used to keep track of the children associated with each
+// variable. It is shared with the host via a scopes response and is an index to
+// the s.variableHandles map, so it can be referenced from a subsequent variables
+// request. A positive reference signals the host that another variables request
+// can be issued to get the elements of the compound variable. As a custom, a zero
+// reference, reminiscent of a zero pointer, is used to indicate that a scalar
+// variable can be "dereferenced" to get its elements (as there are none).
```

Comment:
"is used to indicate that a scalar variable **can** be "dereferenced" to get its elements (as there are none)"

can or cannot?

Reply:
Oops. Done.